### PR TITLE
Feat: Tezos MessageParser & generic state interpretation

### PR DESCRIPTION
**Pull request by gitsimon**

### CODE REVIEW

File: pallets/hyperdrive/src/types.rs [hunk 1]
```diff
@@ -36,3 +43,99 @@ pub enum StateTransmitterUpdate<AccountId, BlockNumber> {
...
+
+pub trait ActionExecutor<AccountId, Extra> {
+    fn execute(action: ParsedAction<AccountId, Extra>) -> DispatchResultWithPostInfo;
+}
+
+impl<AccountId, Extra> ActionExecutor<AccountId, Extra> for () {
+    fn execute(_: ParsedAction<AccountId, Extra>) -> DispatchResultWithPostInfo {
+        Ok(().into())
+    }
+}
```
Comment (RomarQ) [inline pallets/hyperdrive/src/types.rs:141]: `unit` implementation does not make sense here. I'm fine if this gets moved to `mock.rs`
```suggestion
```

File: pallets/hyperdrive/src/tezos.rs [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+                    (to_multi_origin(&origin)?, job_id_sequence),
+                    registration,
+                )
+            }
+        })
+    }
+}
+
+impl<T: Config, I> From<ValidationError> for Error<T, I> {
+    fn from(_: ValidationError) -> Self {
```
Comment (RomarQ) [inline pallets/hyperdrive/src/tezos.rs:66]: The `ValidationError` is lost here. We should at least log it.

File: pallets/hyperdrive/src/tezos.rs [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+}
+
+#[cfg_attr(rustfmt, rustfmt::skip)]
+fn message_schema() -> &'static Micheline {
+    static MESSAGE_SCHEMA: OnceBox<Micheline> = OnceBox::new();
+    MESSAGE_SCHEMA.get_or_init(|| {
+        let schema: Micheline = pair(vec![
+            string(),
+            address(),
+            bytes(),
```
Comment (RomarQ) [inline pallets/hyperdrive/src/tezos.rs:78]:
```suggestion
            // ACTION NAME
            string(),
            // TEZOS ORIGIN
            address(),
            // ACTION PAYLOAD
            bytes(),
```

File: pallets/hyperdrive/src/tezos.rs [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    let unpacked: Micheline = Micheline::unpack(encoded, Some(registration_payload_schema()))
+        .map_err(|e| ValidationError::TezosMicheline(e))?;
+
+    let p: PrimitiveApplication = unpacked.try_into()?;
+    let pair: Pair = p.try_into()?;
+
+    let values = pair.flatten().values;
+    if values.len() != 18 {
+        Err(ValidationError::InvalidMessage)?;
+    }
```
Comment (RomarQ) [inline pallets/hyperdrive/src/tezos.rs:224]: This is already checked by the unpacking.
```suggestion
```

File: pallets/hyperdrive/src/tezos.rs [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+    let unpacked: Micheline = Micheline::unpack(encoded, Some(message_schema()))
+        .map_err(|e| ValidationError::TezosMicheline(e))?;
+
+    let p: PrimitiveApplication = unpacked.try_into()?;
+    let pair: Pair = p.try_into()?;
+
+    let values = pair.flatten().values;
+    if values.len() != 3 {
+        Err(ValidationError::InvalidMessage)?;
+    }
```
Comment (RomarQ) [inline pallets/hyperdrive/src/tezos.rs:102]: This is already checked by the unpacking.
```suggestion
```

File: pallets/hyperdrive/src/tezos.rs [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+}
+
+/// Utility function to parse a tezos [`Sequence`] into a [`Vec`], applying a conversion operation to each item of the sequence.
+fn try_sequence<R, O: Fn(Data) -> Result<R, ValidationError>>(
+    value: Data,
+    op: O,
+) -> Result<Vec<R>, ValidationError> {
+    let s: Sequence = value.try_into()?;
+    s.into_values().into_iter().map(|item| op(item)).collect()
+}
```
Comment (RomarQ) [inline pallets/hyperdrive/src/tezos.rs:493]: I think this is already supported by the rust sdk.

**Review by RomarQ — CHANGES_REQUESTED** [review]

File: pallets/hyperdrive/Cargo.toml [hunk 7]
```diff
@@ -17,6 +17,16 @@ codec = { package = "parity-scale-codec", version = "3.0.0", default-features =
 	"derive",
 ] }
 sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.36" }
+pallet-acurast = { path = "../acurast", default-features = false }
+pallet-acurast-marketplace = { path = "../marketplace", default-features = false}
+xcm = { git = "https://github.com/paritytech/polkadot", default-features = false, branch = "release-v0.9.36" }
+
+# SDKs
+tezos-core = { git = "https://github.com/airgap-it/tezos-rust-sdk", default-features = false, branch = "main" }
+tezos-michelson = { git = "https://github.com/airgap-it/tezos-rust-sdk", default-features = false, branch = "main" }
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/hyperdrive/Cargo.toml]: this works since I migrated tezos libs to no_std

File: pallets/hyperdrive/src/tezos.rs (hunk 2 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/tezos.rs]: we now emit events for error cases (and not fail extrinsic `submit_message`)

File: pallets/hyperdrive/src/tezos.rs (hunk 5 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/tezos.rs:138]: it probably does now, but will it do in all future?

File: pallets/hyperdrive/src/tezos.rs (hunk 4 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/tezos.rs:268]: it probably does now, but will it do in all future?

Comment (RomarQ) [reply pallets/hyperdrive/src/tezos.rs:268]: Yes, the bytes need to match the type.

File: pallets/hyperdrive/src/tezos.rs (hunk 5 above)
Comment (RomarQ) [reply pallets/hyperdrive/src/tezos.rs:138]: Yes, the bytes need to match the type.

**Review by RomarQ — APPROVED** [review]

File: pallets/hyperdrive/src/tezos.rs [hunk 8]
```diff
@@ -0,0 +1,641 @@
...
+                    ),
+                    registration,
+                )
+            }
+        })
+    }
+}
+
+/// Parses an encoded key from Tezos representing a message identifier.
+pub fn unpack_nat(encoded: &[u8]) -> Result<u128, ValidationError> {
```
Comment (gitsimon) [inline pallets/hyperdrive/src/tezos.rs:95]: You could make it generic and name it like that, but the error type returned is specific...so you would have to move to the caller.

**Pull request merged**